googleapiclient/http.py

```diff
@@ -82,22 +89,65 @@ def _retry_request(http, num_retries, req_type, sleep, rand, uri, method, *args,
     resp, content - Response from the http request (may be HTTP 5xx).
   """
   resp = None
+  content = None
   for retry_num in range(num_retries + 1):
     if retry_num > 0:
-      sleep(rand() * 2**retry_num)
+      # Sleep before retrying.
+      sleep_time = rand() * 2 ** retry_num
       logging.warning(
-          'Retry #%d for %s: %s %s%s' % (retry_num, req_type, method, uri,
+          'Sleeping %.2f seconds before retry %d of %d for %s: %s %s%s' %
+          (sleep_time, retry_num, num_retries, req_type, method, uri,
           ', following status: %d' % resp.status if resp else ''))
+      sleep(sleep_time)
 
     try:
       resp, content = http.request(uri, method, *args, **kwargs)
-    except ssl.SSLError:
+    # Retry on SSL errors and socket timeout errors.
+    except (ssl.SSLError, TimeoutError) as ex:
+      # If this is a socket.error, only retry if it's a timeout.
+      if isinstance(ex, socket.error) and not isinstance(ex, ssl.SSLError):
+        try:
+          # errno's contents differ by platform so we have to match by name.
+          if not (socket.errno.errorcode[ex.errno] in
```

Comment:
The difference is that I need to check for `WSAETIMEDOUT` as well, but it isn't defined when not on Windows. I thought it's more readable and more elegant to check it like that instead of handling `AttributeError`.